File: src/main/java/no/novari/flyt/configuration/validation/constraints/UniqueChildrenKeys.java
package no.novari.flyt.configuration.validation.constraints;

import jakarta.validation.Constraint;
import jakarta.validation.Payload;
import java.lang.annotation.Retention;

import static java.lang.annotation.RetentionPolicy.RUNTIME;

@Retention(RUNTIME)
@Constraint(validatedBy = UniqueChildrenKeysValidator.class
)
public @interface UniqueChildrenKeys {

    String DUPLICATE_CHILDREN_KEYS_REF = "duplicateChildrenKeys";

    String message() default "contains duplicate children keys: {" + DUPLICATE_CHILDREN_KEYS_REF + "}";

    Class<?>[] groups() default {};

    Class<? extends Payload>[] payload() default {};

}

File: src/main/kotlin/no/novari/flyt/configuration/validation/constraints/UniqueChildrenKeys.kt
package no.novari.flyt.configuration.validation.constraints

import jakarta.validation.Constraint
import jakarta.validation.Payload
import kotlin.reflect.KClass

@Retention(AnnotationRetention.RUNTIME)
@Target(AnnotationTarget.CLASS)
@Constraint(validatedBy = [UniqueChildrenKeysValidator::class])
annotation class UniqueChildrenKeys(
    val message: String = "contains duplicate children keys: {$DUPLICATE_CHILDREN_KEYS_REF}",
    val groups: Array<KClass<*>> = [],
    val payload: Array<KClass<out Payload>> = [],
) {
    companion object {
        const val DUPLICATE_CHILDREN_KEYS_REF = "duplicateChildrenKeys"
    }
}
